lms/djangoapps/course_blocks/transformers/utils.py

```diff
@@ -18,6 +19,70 @@ def get_field_on_block(block, field_name, default_value=None):
     return default_value
 
 
+def collect_nearest_subsection(
+        block_structure,
+        transformer,
+):
+    """
+    TODO: take another editing pass at this docstring before merge
+    A very specialized (as in, I ought to see if we can generalize this or move it to the grading transformer's file)
+    collect method designed to calculate and store the nearest containing subsection for every block in the course
+    tree. This is done in 2 passes - the first will calculate and store a tuple, 'nearest_sub_dict', for each block
+    in a topological traversal of the tree. This tuple will store the closest containing subsection and its distance
+    from the block. On the second pass, we remove 'nearest_sub_dict' and only store the actual containing subsection,
+    where one exists.
+    """
+    for block_key in block_structure.topological_traversal():
+        if block_key.block_type == 'sequential':
+            block_structure.set_transformer_block_field(
+                block_key,
+                transformer,
+                'nearest_sub_dict',
+                (0, block_key)
+            )
+            continue
+        parent_data = [
+            block_structure.get_transformer_block_field(
+                parent,
+                transformer,
+                'nearest_sub_dict',
+                (BIG_NUMBER, None),
+            ) for parent in block_structure.get_parents(block_key)
+        ]
+        nearest = (BIG_NUMBER, None)
+        for result_pair in parent_data:
+            if result_pair[1] and result_pair[0] < nearest[0]:
+                nearest = result_pair
+        if nearest[1]:
+            nearest = (nearest[0] + 1, nearest[1])
+        block_structure.set_transformer_block_field(
+            block_key,
+            transformer,
+            'nearest_sub_dict',
+            nearest,
+        )
+
+    for block_key in block_structure.topological_traversal():
+        nearest = block_structure.get_transformer_block_field(
+            block_key,
+            transformer,
+            'nearest_sub_dict',
+            (BIG_NUMBER, None),
+        )
+        block_structure.remove_transformer_block_field(
+            block_key,
+            transformer,
+            'nearest_sub_dict'
+        )
+        if nearest[1]:
+            block_structure.set_transformer_block_field(
+                block_key,
+                transformer,
+                'containing_subsection',
+                nearest[1],
+            )
+
+
```

Comment:
Hmm.. it's not necessary to find the "nearest" subsection.  As discussed, we need to find the union of all parent subsections of the block.  The following code does the trick.  In fact, it also works for subsections nested within subsections.

```
def collect_merged_subsection_field(
    block_structure,
    transformer,
    merged_field_name,
):
    """
    Collects a merged set of the locations of all subsection ancestors for each block in the given block structure.
    """
    import itertools
    for block_key in block_structure.topological_traversal():
        # compute merged value of the field from all parents
        parents = block_structure.get_parents(block_key)
        all_parents_merged_value = set(
            itertools.chain.from_iterable([
                block_structure.get_transformer_block_field(
                    parent_key, transformer, merged_field_name, [],
                )
                for parent_key in parents
        ])) if parents else set()

        # set the merged value for this block
        parent_subsections = set(parent for parent in parents if parent.block_type == 'sequential')
        block_structure.set_transformer_block_field(
            block_key,
            transformer,
            merged_field_name,
            all_parents_merged_value | parent_subsections,
        )
```

The grading code can then enumerate through all the ancestor subsections listed in `merged_field_name` and update the grade for each of them.